File: Other problems/aanimals.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
    int n, k;
    cin >> n >> k;

    map<int, set<string>> items;
    for (int i = 0; i < k; i++) {
        int key;
        string value;
        cin >> key >> value;
        items[key].insert(value);
    }

    int m;
    cin >> m;
    vector<string> l(m);
    for (auto &x: l) cin >> x;


    vector<int> path, path2;
    int i_need=0, i_shop=0;

    while (i_need < m and i_shop < n) {
        if (items[i_shop].find(l[i_need]) != items[i_shop].end()) {
            i_need += 1;
            path.push_back(i_shop);
        } else {
            i_shop += 1;
        }
    }

    if (i_need != m) {
        cout << "impossible" << endl;
        return 0;
    }

    i_need = m-1;
    i_shop = n-1;
    while (i_need >= 0 and i_shop >= 0) {
        if (items[i_shop].find(l[i_need]) != items[i_shop].end()) {
            i_need -= 1;
            path2.push_back(i_shop);
        } else {
            i_shop -= 1;
        }
    }

    if (path.size() != path2.size()) {
        cout << "ambiguous" << endl;
        return 0;
    }

    for (int i=0; i<path.size(); i++){
        if (path[i] != path2[path.size()-1-i]) {
            cout << "ambiguous" << endl;
            return 0;
        }
    }

    cout << "unique" << endl;
    return 0;
}

File: Other problems/aanimals.py
from collections import defaultdict

n = int(input())
k = int(input())

items = defaultdict(set)

for _ in range(k):
    s = input().split()
    i = int(s[0])
    x = s[1]

    items[i].add(x)

m = int(input())
l = [input() for _ in range(m)]

i_need = 0
i_shop = 0
path = []
while i_need < m and i_shop < n:
    if l[i_need] in items[i_shop]:
        i_need += 1
        # if not path or path[-1] != i_shop:
        path.append(i_shop)

    else:
        i_shop += 1

if i_need != m:
    print("impossible")
    exit()

i_need = m-1
i_shop = n-1
path2 = []
while i_need >= 0 and i_shop >= 0:
    if l[i_need] in items[i_shop]:
        i_need -= 1
        # if not path2 or path2[-1] != i_shop:
        path2.append(i_shop)

    else:
        i_shop -= 1

# print(path)
# print(path2)
if len(path) == len(path2) and all([path[len(path)-i-1] == path2[i] for i in range(len(path))]):
    print("unique")
else:
    print("ambiguous")




